ThunkLibs/include/common/Host.h

```diff
@@ -75,4 +84,78 @@ struct GuestcallInfo {
 #elif defined(_M_ARM_64)
 #define LOAD_INTERNAL_GUESTPTR_VIA_CUSTOM_ABI(target_variable) \
   asm volatile("mov %0, x11" : "=r" (target_variable))
+#else
+#define LOAD_INTERNAL_GUESTPTR_VIA_CUSTOM_ABI(target_variable) \
+  abort()
 #endif
+
+template<typename>
+struct CallbackUnpack;
+
+template<typename Result, typename... Args>
+struct CallbackUnpack<Result(Args...)> {
+    static Result CallGuestPtr(Args... args) {
+        GuestcallInfo *guestcall;
+        LOAD_INTERNAL_GUESTPTR_VIA_CUSTOM_ABI(guestcall);
+
+        PackedArguments<Result, Args...> packed_args = {
+            args...
+        };
+        guestcall->CallCallback(guestcall->GuestUnpacker, guestcall->GuestTarget, &packed_args);
+        if constexpr (!std::is_void_v<Result>) {
+            return packed_args.rv;
+        }
+    }
+
+    static void ForIndirectCall(void* argsv) {
+        auto args = reinterpret_cast<PackedArguments<Result, Args..., uintptr_t>*>(argsv);
+        constexpr auto CBIndex = sizeof...(Args);
+        uintptr_t cb;
+        static_assert(CBIndex <= 17 || CBIndex == 23);
+        if constexpr(CBIndex == 0) {
```

Comment:
can this be done with a0 being a undefined sized array + alloca?

Reply:
(Replying here since GitHub doesn't have threads in the "main" comment page)

I like the generic PackedArguments implementation, though I'd like to see its impact on compile time before adopting something like it. The current approach is very predictable in this regard due to being simple (if not particularly elegant), whereas the metaprogramming involved here could be more involved in terms of computational expense and symbol pollution (to store the offset for each set of template arguments).